plugins/backend-ekg/src/Cardano/BM/Backend/Prometheus.lhs

```diff
@@ -42,28 +77,58 @@ spawnPrometheus ekg host port = Async.async $
+            rtsMetrics = extractRtsGcMetrics rtsNamespace rtsSamples
+            otherMetrics = extractOtherMetrics otherSamples
+        writeLBS $ A.encode [rtsMetrics, otherMetrics]
         pure ()
-    renderSamples :: [(Text, Value)] -> Builder
-    renderSamples [] = mempty
-    renderSamples samples = mconcat
-        [ case sv of
-            Counter c -> renderNamedValue sk (int64Dec c)
-            Gauge g -> renderNamedValue sk (int64Dec g)
-            Label l -> if isFloat l
-                         then renderNamedValue sk (byteString $ encodeUtf8 l)
-                         else mempty
-            _ -> mempty
-        | (sk,sv) <- samples ]
-    renderNamedValue :: Text -> Builder -> Builder
-    renderNamedValue nm bld =
-        (byteString $ prepareName nm)
-        <> charUtf8 ' '
-        <> bld
-        <> charUtf8 '\n'
-    prepareName nm = encodeUtf8 $ replace " " "_" $ replace "-" "_" $ replace "." "_" nm
-    isFloat v = case double v of
-        Right (_n, "") -> True  -- only floating point number parsed, no leftover
-        _ -> False
 
+    -- rts.gc metrics are always here because they are predefined in ekg-core,
+    -- so we can group them.
+    extractRtsGcMetrics :: Text -> [(Text, Value)] -> MetricsGroup
+    extractRtsGcMetrics ns samples = MetricsGroup
+        { namespace = ns
+        , metrics =
+            [ case sv of
+                Counter c -> intMetric sk c
+                Gauge g   -> intMetric sk g
+                _         -> NoMetric -- rts.gc can contain Counter or Gauge only.
+            | (sk, sv) <- samples
+            ]
+        }
+      where
+        intMetric sk v =
+            Metric { mName  = maybe "" id $ T.stripPrefix (ns <> ".") sk
+                   , mType  = "int" -- All values are Int64.
+                   , mValue = NumberInt (fromIntegral v)
+                   }
+
+    -- We cannot make any assumptions about the format of 'sk' in other samples,
+    -- so group other samples into 'common' group.
+    extractOtherMetrics :: [(Text, Value)] -> MetricsGroup
+    extractOtherMetrics samples = MetricsGroup
+        { namespace = "common"
+        , metrics =
+            [ case sv of
+                Counter c -> mkMetric sk $ NumberInt (fromIntegral c)
+                Gauge g   -> mkMetric sk $ NumberInt (fromIntegral g)
+                Label l   -> case double l of
+                                 Left _       -> NoMetric
+                                 Right (r, _) -> mkMetric sk $ NumberReal r
+                _         -> NoMetric
+            | (sk, sv) <- samples
+            ]
+        }
+      where
+        mkMetric sk number =
+            let (withoutType, typeSuffix) = stripTypeSuffix sk number
+            in Metric { mName = withoutType, mType = typeSuffix, mValue = number }
+        stripTypeSuffix sk number =
+            let types = ["us", "ns", "s", "B", "int", "real"]
+            in case find (\t -> ("." <> t) `T.isSuffixOf` sk) types of
```

Comment:
@denisshevchenko, this still re-traverses the entire string until it finds a match.

The idea is that we do this once, to find a possible suffix, and _then_ we compare it to the set of suffixes.

Reply:
Why is it better?

Reply:
Because you don't retraverse the entire string each time, just to find its suffix.

Reply:
I did similar changes: take suffix as a part after last `.`, and test if it's a part of a list of predetermined set.